beacon-chain/sync/initial-sync/blocks_fetcher.go

```diff
@@ -0,0 +1,400 @@
+type fetchRequestParams struct {
+	ctx   context.Context // if provided, it is used instead of global fetcher's context
+	start uint64          // starting slot
+	count uint64          // how many slots to receive (fetcher may return fewer slots)
+}
+
+// fetchRequestResponse is a combined type to hold results of both successful executions and errors.
+// Valid usage pattern will be to check whether result's `err` is nil, before using `blocks`.
+type fetchRequestResponse struct {
+	params *fetchRequestParams
+	blocks []*eth.SignedBeaconBlock
+	err    error
+	peers  []peer.ID
+}
+
+// newBlocksFetcher creates ready to use fetcher.
+func newBlocksFetcher(cfg *blocksFetcherConfig) *blocksFetcher {
+	rateLimiter := cfg.rateLimiter
+	if rateLimiter == nil {
+		rateLimiter = leakybucket.NewCollector(allowedBlocksPerSecond, allowedBlocksPerSecond, false)
+	}
+
+	ctx := cfg.ctx
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
+	return &blocksFetcher{
+		ctx:                    ctx,
+		headFetcher:            cfg.headFetcher,
+		p2p:                    cfg.p2p,
+		rateLimiter:            rateLimiter,
+		requests:               make(chan *fetchRequestParams),
+		receivedFetchResponses: make(chan *fetchRequestResponse),
+		quit:                   make(chan struct{}),
+	}
+}
+
+// start boots up the fetcher, which starts listening for incoming fetch requests.
+func (f *blocksFetcher) start() {
+	go f.loop()
+}
+
+// stop terminates all fetcher operations.
+func (f *blocksFetcher) stop() {
+	select {
+	case <-f.quit:
+	default:
+		close(f.quit)
+	}
+}
+
+// requestResponses exposes a channel into which fetcher pushes generated request responses.
+func (f *blocksFetcher) requestResponses() <-chan *fetchRequestResponse {
+	return f.receivedFetchResponses
+}
+
+// loop is a main fetcher loop, listens for incoming requests/cancellations, forwards outgoing responses.
+func (f *blocksFetcher) loop() {
+	defer close(f.receivedFetchResponses)
+
+	for {
+		select {
+		case <-f.ctx.Done():
+			// Upstream context is done.
+			err := errors.New("upstream context canceled")
+			log.WithError(err).Warn("Block fetch loop closed unexpectedly")
+			f.receivedFetchResponses <- &fetchRequestResponse{
+				err: err,
+			}
+			f.stop()
+			return
+		case <-f.quit:
+			// Terminating abort all operations.
+			log.Debug("Blocks fetcher received a stop request")
+			return
+		case req := <-f.requests:
+			if req.ctx == nil {
+				req.ctx = f.ctx
+			}
+			go f.handleRequest(req)
+		}
+	}
+}
+
+// scheduleRequest adds request to incoming queue.
+func (f *blocksFetcher) scheduleRequest(ctx context.Context, start, count uint64) {
+	select {
+	case <-f.quit:
+		return
+	default:
+		f.requests <- &fetchRequestParams{
+			ctx:   ctx,
+			start: start,
+			count: count,
+		}
+	}
+}
+
+// handleRequest parses fetch request and forwards it to response builder.
+func (f *blocksFetcher) handleRequest(req *fetchRequestParams) {
```

Comment:
Can we add jaeger spans to this function? Similar to how we do elsewhere in the repo such as https://github.com/prysmaticlabs/prysm/blob/c09ae21ab0041da1313f96f3b6f81995761b2e89/beacon-chain/blockchain/head.go#L26. Would be nice to stick to the pattern of ctx as first argument to most functions for consistency. It also allows you to do easier unit testing as you can pass in a context you can easily cancel in your tests.